### DISCUSSION TITLE
Migration to independent platform for provision of source code

### DISCUSSION THREAD

**Discussion by linux-lukas** [streetcomplete/StreetComplete#4268] [General]
### A. Problem / Goal
Since the purchase of GitHub by Microsoft in 2018, a dependence on the BigTech corporation can no longer be denied.

On the one hand, I can understand why GitHub was chosen as the platform for making source code available: "Everyone is here".

On the other hand, I see the danger of a vendor-lockin effect and that open source projects become centrally dependent on Microsoft. In my eyes, this is very dangerous for free and open source software and hardware projects.

In the medium and long term, the goal would be to become independent of GitHub and thus of Microsoft. The Gitea-based Codeberg project of Codeberg e. V. in Berlin would be a good choice here.

There are also (legal) problems with compliance with the licence of GitHub functions, such as the co-pilot.

### B. Solution
My considered solution to the problem described in A. would be the following:

A user of this open source project creates a user account on https://codeberg.org/
If necessary: This user creates an organisation for the project.
A "personal access token" is created on the GitHub account, which has appropriate rights to the organisation repositories, using the developer options in the settings.
all repositories would be migrated with this access token into the ownership of the organisation created in step two.
Regarding step four, there is an entry in the documentation of Codeberg: https://docs.codeberg.org/advanced/migrating-repos/

### C. Alternatives
A possible alternative would be to perform the first three steps as described in B. A possible alternative would be to perform the first three steps as described in B., and modify the fourth step to include a mirror of GitHub. So that all issues and such that would be created in the GitHub repository would be transferred to the Codeberg repository.

### D. Responsibilities
I would see the responsibility in the owners of the repository and, if necessary, additional project participants.

### E. Other
Basically, a look at the documentation of Codeberg is not unwise: https://docs.codeberg.org/

Should it be necessary to manage repositories in organisations, this is also possible under Codeberg, see: https://docs.codeberg.org/collaborating/create-organization/

Regarding licensing there is a page in the documentation of Codeberg: https://docs.codeberg.org/getting-started/licensing/

Funfact 1: The Pixelfed maintainer has written on Mastodon that he is considering migrating the Pixelfed project away from GitHub, to Codeberg: https://mastodon.social/@dansup/10869007146309141

Funfact 2: Aral had migrated - also on Mastodon - to Codeberg for his GitHub projects as well as those of his Foundation, which he organised together with another person, Small Technology, or is still migrating: https://mastodon.ar.al/@aral/108651118032076133

**Comment by HolgerJeromin** [comment]
F: Risk

Less bug reports, discussions as peoples do not want to create an account for another platform they never heard before.

**Reply by matkoniecz** [reply]
Also, with Codeberg there is higher risk of service disappearing, changing rules, losing data and otherwise causing problems.

**Reply by linux-lukas** [reply]
On the one hand, I would agree with them right now that it could happen in the short term. On the other hand, there are currently (very) active efforts to implement a kind of federation (like the Fediverse) in the web-based git versioning and source code delivery platform gitea. 

Example: so that source code can be on codeberg, an issue can be created with a gitlab account (assuming GitLab would implement this capability even if gitea implemented it).

And it seems that steady progress is being made towards this goal, see the following Mastodon post: 
https://social.exozy.me/@ta180m/108631221939677386

**Reply by linux-lukas** [reply]
> Also, with Codeberg there is higher risk of service disappearing, changing rules, losing data and otherwise causing problems.

Yes, but I think these points would also apply to GitHub or GitLab.

**Comment by matkoniecz** [comment]
I understand your position, and I strongly prefer open source and Codeberg seems more promising than Gitlab from a very quick view.

But I am not free software maximalist, for example I am not treating "no binary blobs in the internal software of CPU" as mandatory and I am not planning to spend effort on such change.

**Reply by linux-lukas** [reply]
I can understand your point, because this involves a lot of effort, of course, and this should not be the task of one project, but rather an association (e.g. via the FSFE) must become active, i.e. several projects together.

**Reply by matkoniecz** [reply]
BTW, thanks for making me aware of Codeberg. I will likely use it for less important projects and ones where external contributions are less wanted.
